src/Workspaces/Core/Portable/Workspace/Host/DocumentService/IDocumentServiceProvider.cs

```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    internal interface IDocumentServiceProvider
```

Comment:
❓ Why not abstract class `HostDocumentServices`, similar to how workspace and language services are handled?

Reply:
I opened - https://github.com/dotnet/roslyn/issues/30937 for the discussion.

but for now, it is not following that pattern since we are not sure whether it is host services or not (services that host who owns the Roslyn provide).

rather it is a service external component inject to us like IWorkspaceServiceFactory or ILanguageServiceFactory.

Reply:
we can change name to IDocumentServiceFactory if we want to. Factor or provider, I don't care much, just let's pick one from design meeting and let me know and I will change name to it.

Reply:
if I have put this in MS.CA.Host is the problem, then sure we can move to other namespace once we decide to what that is.

I have put it here since all interface external component can implements are under this namespace.

basically, roslyn has 2 different type of interfaces.

1. interfaces Roslyn (including F#, Typescript and etc) is expected to implement and external people (Features) consume the service thorough the interface. we do this for internal services (IFormattingServices, ICodeFixServices and etc), and we make those a static type when we expose it to public (Formatter, QuickInfo and etc)

2. interfaces Roslyn (Workspace) consumes that we have default implementation but external people can provide their own host specific implementation. we call this kind of interface as Host services such as temporary storage services, persistent services, metadata provider, xml document provider and etc.

so, it is not really Host services since we don't expect like VS for Mac Workspace or Any Code Workspace or Console Workspace to provide their own implementation but, it is not the (1) either since we are not supposed to implement the services for razor either.

anyway, just I have put it in Host namespace. we can change to whatever namespace we want if we find better place to put it. we can create like MS.CA.External or something.